astropy/units/quantity.py

```diff
@@ -576,40 +587,69 @@ def _new_view(self, obj, unit=None):
-            to self._unit.
+            subclass, and explicitly assigned to the view if given.
+            If not given, the subclass and unit will be that of ``self``.
 
         Returns
         -------
         view : Quantity subclass
         """
-        # python and numpy scalars cannot be viewed as arrays and thus not as
-        # Quantity either; turn them into zero-dimensional arrays
-        # (These are turned back into scalar in `.value`)
-        if not isinstance(obj, np.ndarray):
-            obj = np.array(obj)
-
+        # Determine the unit and quantity subclass that we need for the view.
         if unit is None:
-            subclass = self.__class__
+            unit = self.unit
+            quantity_subclass = self.__class__
         else:
             unit = Unit(unit)
-            subclass, subok = self.__quantity_subclass__(unit)
+            quantity_subclass, subok = self.__quantity_subclass__(unit)
             if subok:
-                subclass = self.__class__
+                quantity_subclass = self.__class__
+
+        # We only want to propagate information from ``self`` to our new view,
+        # so obj should be a regular array.  By using ``np.array``, we also
+        # convert python and numpy scalars, which cannot be viewed as arrays
+        # and thus not as Quantity either, to zero-dimensional arrays.
+        # (These are turned back into scalar in `.value`)
+        if obj is None:
+            obj = self.view(np.ndarray)
+        else:
+            obj = np.array(obj, copy=False)
 
-        view = obj.view(subclass)
+        # Take the view, set the unit, and update possible other properties
+        # such as ``info``, ``wrap_angle`` in `Longitude`, etc.
+        view = obj.view(quantity_subclass)
+        view._set_unit(unit)
         view.__array_finalize__(self)
-        if unit is not None:
-            view._unit = unit
         return view
 
+    def _set_unit(self, unit):
+        """Set the unit.
+
+        This is used anywhere the unit is set or modified, i.e., in the
+        initilizer, in ``__imul__`` and ``__itruediv__`` for in-place
+        multiplication and division by another unit, as well as in
+        ``__array_finalize__`` for wrapping up views.  For Quantity, it just
+        sets the unit, but subclasses can override it to check that, e.g.,
+        a unit is consistent.  It should return ``self`` after modification.
+        """
+        if not isinstance(unit, UnitBase):
+            # Trying to go through a string ensures that, e.g., Magnitudes with
+            # dimensionless physical unit become Quantity with units of mag.
+            unit = Unit(str(unit), parse_strict='silent')
+            if not isinstance(unit, UnitBase):
+                raise UnitTypeError(
+                    "{0} instances require {1} units, not {2} instances."
+                    .format(type(self).__name__, UnitBase, type(unit)))
+
+        self._unit = unit
+        return self
```

Comment:
[very minor question] 
I'm curious why you return `self` here -- since this will always be called as `self._set_unit(...)` the instance should be on hand already, right?